File: LIS.py
n = int(input())
a = [0] + list(map(int, input().split()))
L = [0]
for i in range(1, n+1):
    if a[i] > L[-1]: L.append(a[i])
    else:
        low, high = 0, len(L) - 1
        p = 0
while low <= high:
            mid = (low + high) // 2
if L[mid] >= a[i]:
                high = mid - 1
else:
                p = mid
                low = mid + 1
        L[p+1] = a[i]
print(len(L)-1)

File: LIS.cpp
#include <bits/stdc++.h>
using namespace std;
int n, x, a[100005];
vector <int> l;
int main(){
    cin >> n;
    l.push_back(0);
    for (int i = 1; i <= n; ++i) cin >> a[i];
    for (int i = 1; i <= n; ++i){
        if (a[i] > l[l.size()-1]) l.push_back(a[i]);
        else{
            int low = 0;
            int high = l.size()-1;
            int p = 0;
            while (low <= high){
                int mid = (low + high) / 2;
                if (l[mid] >= a[i]) high = mid - 1;
                else{
                    low = mid + 1;
                    p = mid;
                }
            }
            l[p+1] = a[i];
        }
    }
    cout << l.size()-1;
}